designs/messaging/messaging-framework.md

```diff
@@ -0,0 +1,413 @@
+2. FAQs
+3. Questions for the community
+
+
+## Framework Components
+
+When systems use messaging for distributing work, there are often multiple actors in the system with each performing a different part of the message processing. For example, a commerce application could use an ASP.NET Core application to publish messages when orders are taken. Backend processes, like an ECS service or Lambda function, could be responsible for consuming the messages and performing the actual work.
+
+### Configuration
+Modern .NET heavily uses dependency injection (DI). AWS SDK for .NET integrates with the .NET DI framework using the **[.NET Dependency Injection Extensions for AWS SDK for .NET](https://docs.aws.amazon.com/sdk-for-net/v3/developer-guide/net-dg-config-netcore.html#net-core-dependency-injection)**. Many of AWS’s other high level libraries integrate with DI as well.
+
+The message processing framework will be configured through the DI framework as well. The main configuration points are:
+
+- Mapping .NET objects to which topics, queues or event bus to publish to (Publish points)
+- Mapping .NET objects to handlers to process messages (Subscriber points)
+- Concurrency and message pump settings
+
+#### Configuration mapping
+For both publishing and subscribing there are 3 important pieces of information to configure.
+
+- The .NET type to use as the container of the message
+- The AWS resource identifier (queue URLs, topic ARNs, event bus ARNs)
+- The message type that will be set in the message envelope when creating the message. If this is not set then the full .NET type name will be used.
+
+A .NET type can only be registered once for publishing or subscribing. When a .NET object is being published the AWS resource and message type name will be retrieved from the mapping based on the .NET type. The .NET object will be serialized along with its message envelope.
+
+When reading messages from queues the message type will be pulled from the message envelope. The message type will be used to deserialize the message to the corresponding .NET type. The handler class that has been mapped to the .NET type will be invoked, passing in the .NET object with its message envelope for handlers that need extra metadata.
+
+#### Configuring a publisher
+An ASP.NET Core application could be a publisher. A typical ASP.NET Core application has a `ConfigureServices` method to configure the DI. In this method the application would configure the types of messages it will send. The example below shows how to register .NET objects to publishing endpoints.
+
+```csharp
+public IConfiguration Configuration { get; }
+
+public void ConfigureServices(IServiceCollection services)
+{
+    // Pull out physical AWS resource ids from .NET Configuration framework
+    var awsResourceIds = new AWSResourceIds();
+    Configuration.GetSection(AWSResourceIds.SectionName).Bind(awsResourceIds);
+            
+    services.AddAWSMessageBus(builder =>
```

Comment:
Have you considered how to deal with serialization options? Our events might have polymorphic properties on them requiring a suitable STJ converter, there should be a way to either tell this library to use an already registered serializer, or pass in a `System.Text.Json.JsonSerializerOptions`.